[![stars](https://img.shields.io/github/stars/BGI-Qingdao/SpaGRN?logo=GitHub&color=yellow)](https://img.shields.io/github/stars/BGI-Qingdao/SpaGRN) 
[![docs](https://img.shields.io/static/v1?label=docs&message=spagrn&color=green)](https://spagrn.readthedocs.io/en/latest/index.html)


## **SpaGRN**
A comprehensive tool to infer TF-centered, spatial gene regulatory networks for the spatially resolved transcriptomics (SRT) data.

SpaGRN is an open-source Python package for inferring gene regulatory networks (GRNs) based on spatial gene expression data using GPLv3 license. The model takes into account the spatial proximity of genes and TF binding to infer their regulatory relationships. The package is particularly useful for analyzing spatially resolved gene expression data.
This approach can be applied to various types of spatial transcriptomics data, such as Stereo-seq, Seq-Scope, Pixel-seq, Slide-seq, MERFISH, STARmap, CosMx, ST, 10x Visium, DBiT-seq, and others. Notably, we are still working on the improvement of performance and calculation efficiency.


RRID: SCR_023451

[**Installation**](https://spagrn.readthedocs.io/en/latest/content/00_Installation.html) **-** 
[**Quick start**](https://spagrn.readthedocs.io/en/latest/content/01_Basic_Usage.html) **-** 
[**Tutorial**](https://spagrn.readthedocs.io/en/latest/Tutorials/index.html) **-** 
[**Documentation**](https://spagrn.readthedocs.io/en/latest/index.html) 


<img src="docs/source/_static/mainpipeline.BMP" width="100%" height="100%">

[comment]: <> (![SpaGRN]&#40;./docs/source/_static/mainpipeline.BMP&#41;)

## Discussion 
Please use GitHub [issues](https://github.com/BGI-Qingdao/SpaGRN/issues) tracker for reports and discussions of:
 - Bug reports
 - Document and data issues
 - Feature requirements
 - ...

## Contribution 
**SpaGRN** is in a stage of rapid development so that we will carefully consider all aspects of your proposal. We hope future input will be given by both users and developers.


## File tree
To enhance clarity and prevent any potential confusion, we have organized this GitHub repository into a structured **file tree**, complete with detailed annotation for each code segment, tailored to their respective purposes.

```

```