References
===========

Carpenter, A.E. et al. CellProfiler: image analysis software for identifying and quantifying cell phenotypes. Genome Biol 7, R100 (2006).

Stuart, T. et al. Comprehensive Integration of Single-Cell Data. Cell 177, 1888-1902 e1821 (2019).

Fedorov, A. et al. 3D Slicer as an image computing platform for the Quantitative Imaging Network. Magn Reson Imaging 30, 1323-1341 (2012).

Cignoni, P. et al. in Eurographics Italian Chapter Conference. (eds. V. Scarano, R.D. Chiara & U. Erra) 129-136 (Eurographics, 2008).

Malandain, G., Bardinet, E., Nelissen, K. & Vanduffel, W. Fusion of autoradiographs with an MR volume using 2-D and 3-D linear transformations. Neuroimage 23, 111-127 (2004).

Jin, S. et al. Inference and analysis of cell-cell communication using CellChat. Nat Commun 12, 1088 (2021).

Cao, J. et al. The single-cell transcriptional landscape of mammalian organogenesis. Nature 566, 496-502 (2019).

La Manno, G. et al. RNA velocity of single cells. Nature 560, 494-498 (2018).

Qiu, X. et al. Mapping transcriptomic vector fields of single cells. Cell 185, 690-711 e645 (2022).

Qiu, X. et al. Single-cell mRNA quantification and differential analysis with Census. Nat Methods 14, 309-315 (2017).

Dann, E., Henderson, N.C., Teichmann, S.A., Morgan, M.D. & Marioni, J.C. Differential abundance testing on single-cell data using k-nearest neighbor graphs. Nat Biotechnol 40, 245-253 (2022).

Miao, Z., Deng, K., Wang, X. & Zhang, X. DEsingle for detecting three types of differential expression in single-cell RNA-seq data. Bioinformatics 34, 3223-3224 (2018).
