References
===========

Yao Li, Xiaobin Liu, Lidong Guo, Kai Han, Shuangsang Fang, Xinjiang Wan, Xun Xu, Guangyi Fan, Mengyang Xu (2023). SpaGRN: investigating spatially informed regulatory paths for spatially resolved transcriptomics data. bioRxiv 10.1101/2023.11.19.567673.

Pratapa, A., Jalihal, A.P., Law, J.N., Bharadwaj, A., and Murali, T.M. (2020). Benchmarking algorithms for gene regulatory network inference from single-cell transcriptomic data. Nat Methods 17, 147-154. 10.1038/s41592-019-0690-6.

Badia, I.M.P., Wessels, L., Müller-Dott, S., Trimbour, R., Ramirez Flores, R.O., Argelaguet, R., and Saez-Rodriguez, J. (2023). Gene regulatory network inference in the era of single-cell multi-omics. Nat Rev Genet. 10.1038/s41576-023-00618-5.

Huynh-Thu, V.A., Irrthum, A., Wehenkel, L., and Geurts, P. (2010). Inferring regulatory networks from expression data using tree-based methods. PLoS One 5. 10.1371/journal.pone.0012776.

Moerman, T., Aibar Santos, S., Bravo González-Blas, C., Simm, J., Moreau, Y., Aerts, J., and Aerts, S. (2019). GRNBoost2 and Arboreto: efficient and scalable inference of gene regulatory networks. Bioinformatics 35, 2159-2161. 10.1093/bioinformatics/bty916

Aibar, S., González-Blas, C.B., Moerman, T., Huynh-Thu, V.A., Imrichova, H., Hulselmans, G., Rambow, F., Marine, J.-C., Geurts, P., Aerts, J., et al. (2017). SCENIC: single-cell regulatory network inference and clustering. Nature Methods 14, 1083-1086. 10.1038/nmeth.4463.

Bravo González-Blas, C., De Winter, S., Hulselmans, G., Hecker, N., Matetovici, I., Christiaens, V., Poovathingal, S., Wouters, J., Aibar, S., and Aerts, S. (2023). SCENIC+: single-cell multiomic inference of enhancers and gene regulatory networks. Nat Methods. 10.1038/s41592-023-01938-4.

Van de Sande, B., Flerin, C., Davie, K., De Waegeneer, M., Hulselmans, G., Aibar, S., Seurinck, R., Saelens, W., Cannoodt, R., Rouchon, Q., et al. (2020). A scalable SCENIC workflow for single-cell gene regulatory network analysis. Nature Protocols 15, 2247-2276. 10.1038/s41596-020-0336-2.

Chen, A., Liao, S., Cheng, M., Ma, K., Wu, L., Lai, Y., Qiu, X., Yang, J., Xu, J., Hao, S., et al. (2022). Spatiotemporal transcriptomic atlas of mouse organogenesis using DNA nanoball-patterned arrays. Cell 185, 1777-1792.e1721. 10.1016/j.cell.2022.04.003.

DeTomaso, D., and Yosef, N. (2021). Hotspot identifies informative gene modules across modalities of single-cell genomics. Cell Syst 12, 446-456.e449. 10.1016/j.cels.2021.04.005

Guilliams, M., Bonnardel, J., Haest, B., Vanderborght, B., Wagner, C., Remmerie, A., Bujko, A., Martens, L., Thoné, T., Browaeys, R., et al. (2022). Spatial proteogenomics reveals distinct and evolutionarily conserved hepatic macrophage niches. Cell 185, 379-396.e338.10.1016/j.cell.2021.12.018.

Zhang, Y., Liu, T., Hu, X., Wang, M., Wang, J., Zou, B., Tan, P., Cui, T., Dou, Y., Ning, L., et al. (2021). CellCall: integrating paired ligand receptor and transcription factor activities for cell-cell communication. Nucleic Acids Res 49, 8520-8534. 10.1093/nar/gkab638.

Kleino, I., Frolovaitė, P., Suomi, T., and Elo, L.L. (2022). Computational solutions for spatial transcriptomics. Computational and Structural Biotechnology Journal 20, 4870-4884. 10.1016/j.csbj.2022.08.043.

Dimitrov, D., Türei, D., Garrido-Rodriguez, M., Burmedi, P.L., Nagai, J.S., Boys, C., Ramirez Flores, R.O., Kim, H., Szalai, B., Costa, I.G., et al. (2022). Comparison of methods and resources for cell-cell communication inference from single-cell RNA-Seq data. Nat Commun 13, 3224. 10.1038/s41467-022-30755-0.

Guo, L., Li, Y., Qi, Y., Huang, Z., Han, K., Liu, X., Liu, X., Xu, M., and Fan, G. (2023). VT3D: a visualization toolbox for 3D transcriptomic data. J Genet Genomics 50, 713-719. 10.1016/j.jgg.2023.04.001.

Moran, P.A.P. (1950). Notes on Continuous Stochastic Phenomena. Biometrika 37, 17-23. 10.2307/2332142.

Geary, R.C. (1954). The Contiguity Ratio and Statistical Mapping. The Incorporated Statistician 5, 115-146. 10.2307/2986645.

Getis, A., and Ord, J.K. (1992). The Analysis of Spatial Association by Use of Distance Statistics. Geographical Analysis 24, 189-206. https://doi.org/10.1111/j.1538-4632.1992.tb00261.x.

van der Maaten, L., and Hinton, G. (2008). Visualizing Data using t-SNE. Journal of Machine Learning Research 9, 2579–2605-2579–2605.

Hao, M., Hua, K., and Zhang, X. (2021). SOMDE: a scalable method for identifying spatially variable genes with self-organizing map. Bioinformatics 37, 4392-4398. 10.1093/bioinformatics/btab471.

Li, Z., Wang, T., Liu, P., and Huang, Y. (2023). SpatialDM for rapid identification of spatially co-expressed ligand-receptor and revealing cell-cell communication patterns. Nat Commun 14, 3995.10.1038/s41467-023-39608-w.

Wartenberg, D. (1985). Multivariate Spatial Correlation: A Method for Exploratory Geographical Analysis. Geographical Analysis 17, 263-283. https://doi.org/10.1111/j.1538-4632.1985.tb00849.x.

Browaeys, R., Saelens, W., and Saeys, Y. (2020). NicheNet: modeling intercellular communication by linking ligands to target genes. Nat Methods 17, 159-162. 10.1038/s41592-019-0667-5.