.. SpaGRN manual documentation master file, created by
   sphinx-quickstart on Sun Dec 22 23:21:00 2024.
   You can adapt this file completely to your liking, but it should at least
   contain the root `toctree` directive.

.. =====================
.. Document Title
.. =====================

.. First level
.. -----------

.. Second level
.. ++++++++++++

.. Third level
.. ************

.. Fourth level
.. ~~~~~~~~~~~~


|stars|  |docs| 

SpaGRN
========================================================
.. rubric:: a comprehensive tool to infer TF-centred, spatial gene regulatory networks for the spatially resolved transcriptomic data.

**SpaGRN** is an open-source Python package for inferring gene regulatory networks (GRNs) based on spatial gene expression data using GPLv3 license. The model takes into account the spatial proximity of genes and TF binding to infer their regulatory relationships. The package is particularly useful for analyzing spatially resolved gene expression data.
This approach can be applied to various types of spatial transcriptomics data, such as Stereo-seq, Seq-Scope, Pixel-seq, Slide-seq, MERFISH, STARmap, CosMx, ST, 10x Visium, DBiT-seq, and others. Notably, we are still working on the improvement of performance and calculation efficiency.



* Get quickly started by browsing :ref:`/content/00_Installation.rst`, :ref:`/content/01_Basic_Usage.rst` or :ref:`/tutorials/index.rst`.
* Open to discuss and provide feedback on `Github <https://github.com/BGI-Qingdao/SpaGRN>`_.


Highlights
------------

Conceptual Advances:
++++++++++++++++++++++++

(1) In the context of spatial transcriptomics (SRT) data, SpaGRN introduces a novel statistical approach leveraging spatial information to precisely identify
cell type or functional domain-specific gene regulatory networks (GRNs) that determine the structure and function of localized cells. Instead of relying only
on Pearson's or Spearman's correlation between gene expressions within the same cell, SpaGRN focuses on the spatially aggregated regulations, which
theoretically outperforms existing scRNA-seq-based GRN inference tools, such as pySCENIC.

(2) SpaGRN employs motif enrichment analysis to transform bidirectional spatial co-expressions into directional causal relationships. In this way, SpaGRN
reduces a number of false-positive transcription factor (TF)-target associations, as a common limitation in gene co-expression module inference
methods, such as Hotspot, CellTrek, and SpaceX. These methods might incorrectly identify the co-expression of non-TF genes, which do not represent
any regulatory relationships and should not be included in the inferred regulons.

(3) SpaGRN further incorporates spatial adjacency information to explore the influence of cell-cell interactions on the intracellular regulatory events through
the identification of spatially co-expressed receptors and TFs, allowing for a more holistic view of the intercellular interactions to intracellular dynamics
driving cellular regulations within the spatial context. This functionality sets SpaGRN apart from other approaches that solely focus on intracellular
regulations, including all the competing tools mentioned above.



Methodological Innovations:
++++++++++++++++++++++++++++++

To fully exploit the information of spatial co-expression, regulatory direction, and receptor-involved influence, we have designed a series of algorithms to accommodate
the information integration within the SpaGRNâ€™s framework, and applicable to various types of input generated by different SRT platforms. Specifically, the original
version of SpaGRN consists of four modules, including identifying genes with spatial patterns, determining pairwise spatial co-expression correlations for identified gene
pairs, excluding false-positive correlations through motif enrichment analysis, and linking potential receptors to directional regulons. For instance, to capture the
localized co-expressions and reduce the influence of high-order neighbors, SpaGRN incorporates a spatial kernel with decay weights in the first two modules.

To further improve the methodological capabilities and robustness, SpaGRN has integrated five spatial autocorrelation algorithms and three correlation algorithms to precisely
detect both global gradients and local heterogeneity using ensemble learning in the revised version of SpaGRN. Notably, we have introduced the derivation of the
bivariate Geary's C method for the first time.



Workflow
----------

.. image:: ./_static/mainpipeline.BMP
    :alt: Title figure
    :width: 700px
    :align: center


.. toctree::
    :titlesonly:
    :maxdepth: 4
    :hidden:

    content/00_Installation
    content/01_Basic_Usage
    Tutorials/index
    content/03_References
    


.. |docs| image:: https://img.shields.io/static/v1?label=docs&message=spagrn&color=green
    :target: https://spagrn.readthedocs.io/en/latest/index.html
    :alt: docs

.. |stars| image:: https://img.shields.io/github/stars/BGI-Qingdao/SpaGRN?logo=GitHub&color=yellow
    :target: https://github.com/BGI-Qingdao/SpaGRN
    :alt: stars
